signer/README.md

```diff
@@ -0,0 +1,82 @@
+# Mina signer
+
+This crate provides an API and framework for Mina signing.  It follows the algorithm outlined in the [Mina Signature Specification](https://github.com/MinaProtocol/mina/blob/develop/docs/specs/signatures/description.md).
+
+## Simple interface
+
+The simple interface uses the default signer configuration compatible with mainnet and testnet transaction signatures.
+
+```rust
+use rand;
+use mina_signer::{NetworkId, Keypair, Signer};
+use mina_signer::NetworkId;
+
+let mut ctx = mina_signer::create(NetworkId::TESTNET);
+let sig = ctx.sign(key_pair, transaction);
+
+assert_eq!(ctx.verify(sig, key_pair.public, transaction), true);
+```
+
+## Advanced interface
+
+The advanced interface allows specification of an alternative cryptographic sponge and parameters, for example, in order to create signatures that can be verified more efficiently using the Kimchi proof system.
+
+```rust
+use rand;
+use mina_signer::{NetworkId, Keypair, Signer};
+use oracle::{pasta, poseidon};
+
+let mut ctx = mina_signer::custom::<poseidon::PlonkSpongeConstants5W>(
+    pasta::fp5::params(),
+    NetworkId::TESTNET,
+);
+
+let sig = ctx.sign(key_pair, transaction);
+assert_eq!(ctx.verify(sig, key_pair.public, transaction), true);
+```
+
+## Framework
+
+The framework allows you to easily define a new signature type simply by implementing the `Hashable` and `Signable` traits.
+
+For example, if you wanted to create Mina signatures for a `Foo` structure you would do the following.
+
+```rust
+use mina_signer::{Hashable, NetworkId, ROInput, Signable};
+
+#[derive(Clone, Copy)]
+struct Foo {
+    foo: u32,
+    bar: u64,
+}
+
+impl Hashable for Foo {
+    fn to_roinput(self) -> ROInput {
+        let mut roi = ROInput::new();
+
+        roi.append_u32(self.foo);
+        roi.append_u64(self.bar);
+
+        roi
```

Comment:
did you create an issue to use serde for that :o ? This manual redefinition of the struct really scares me as we change these structs all the time

Reply:
Yes.  https://github.com/o1-labs/proof-systems/issues/241

Agreed!

For now this should be OK because for tx signatures it doesn't change.  For backward compatibility, each version of tx/signatures is a new thing (i.e. new tx version and structure), rather than updating the old one.

Anyway, it's way better to do it automatically and I will try to do this with subsequent ticket 241 :-)